### PR TITLE
Group weekly todo

### PR DESCRIPTION
**Pull request by Adhikram**
## Weekly Todo

[Notion Task Link](https://www.notion.so/Groups-Weekly-Todo-6e9dfe9d1a924ce3a6a236cb9fa3b272)

#### Branches for testing

Frontend: `feat/new-feature`

Backend: group_weekly_todo

#### Linked pull requests

[Frontend](https://github.com/devs-nest/devsnest-frontend/pull/)

### CODE REVIEW

File: devsnest/app/controllers/api/v1/weekly_todo_controller.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  module V1
+    # Controller For Weekly Todo
+    class WeeklyTodoController < ApplicationController
+      include JSONAPI::ActsAsResourceController
+      before_action :check_group_id, only: %i[index]
+      before_action :weekly_status, only: %i[index]
+      before_action :authorize_update, only: %i[update]
+
+      def context
+        { user: @current_user,
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/weekly_todo_controller.rb:13]: rubocop

**Comment by vedansh** [comment]
@sam0000-cmd can you review this, and re check all permissions for this

File: devsnest/app/controllers/api/v1/weekly_todo_controller.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+          params: params }
+      end
+
+      def check_group_id
+        group_id = GroupMember.find_by(user_id: @current_user).group_id
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/weekly_todo_controller.rb:18]: what about admin, batch_leader

File: devsnest/app/controllers/api/v1/weekly_todo_controller.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        render_unauthorized if group_id != params[:group_id].to_i
+      end
+
+      def streak
+        streak = []
+        sheet = WeeklyTodo.where(group_id: params[:id])
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/weekly_todo_controller.rb:24]: what is sheet, it should be plural?, name it better

File: devsnest/app/controllers/api/v1/weekly_todo_controller.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        return render_error('No content') if sheet.nil?
+
+        start_date = sheet.first.creation_week
+        end_date = Date.today.at_beginning_of_week
+        while start_date <= end_date
+          sub = sheet.find_by(creation_week: start_date.at_beginning_of_week)
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/weekly_todo_controller.rb:30]: Use one call to find all data points

File: devsnest/app/controllers/api/v1/weekly_todo_controller.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+          streak.push({ start_date.to_s => sub.present? && sub.sheet_filled ? true : false })
+          start_date += 7
+
+        end
+
+        render json: streak
+      end
+
+      def show
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/weekly_todo_controller.rb:39]: show function should be in resource

File: devsnest/app/controllers/api/v1/weekly_todo_controller.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        WeeklyTodo.where(group_id: params[:group_id], creation_week: Date.parse(params[:date]).at_beginning_of_week)
+      end
+
+      def authorize_update
+        sheet = WeeklyTodo.find(params[:data][:id])
+        render_error('No content') if sheet.nil?
+        curr_group = Group.find(sheet.group_id)
+        render_unauthorized if (curr_group.owner_id != @current_user.id) && (curr_group.co_owner_id.present? && curr_group.co_owner_id != @current_user.id)
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/weekly_todo_controller.rb:47]: permissions function already written in group

File: devsnest/app/resources/api/v1/weekly_todo_resource.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      attributes :group_id, :sheet_filled, :creation_week, :most_active, :most_helpful, :moral_status, :obstacles, :feedback, :todo_list
+      attributes :end_week
+
+      def end_week
+        creation_week + 6.days
+      end
+
+      def creatable_fields(context)
+        super - %i[creation_week]
+      end
```
Comment (sam0000-cmd) [inline devsnest/app/resources/api/v1/weekly_todo_resource.rb:16]: is this needed??

File: devsnest/app/models/weekly_todo.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+# Weekly Todo Resourses
+class WeeklyTodo < ApplicationRecord
+  belongs_to :group, foreign_key: 'group_id'
+  validates :group_id, uniqueness: { scope: :creation_week }
+  def self.create_todo(group_id, date)
+    WeeklyTodo.create(group_id: group_id, creation_week: Date.parse(date).at_beginning_of_week)
+  end
```
Comment (sam0000-cmd) [inline devsnest/app/models/weekly_todo.rb:9]: can we do this in controller

File: devsnest/app/controllers/api/v1/weekly_todo_controller.rb [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+      before_action :weekly_status, only: %i[index]
+
+      def context
+        { user: @current_user,
+          params: params }
+      end
+
+      def streak
+        streak = WeeklyTodo.where(group_id: params[:id]).order('creation_week').pluck('creation_week,sheet_filled')
+        unless streak.empty?
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/weekly_todo_controller.rb:18]: return json: [] if streak.empty?

File: devsnest/app/controllers/api/v1/weekly_todo_controller.rb [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        end
+
+        render_forbidden
+      end
+
+      def weekly_status
+        todo = WeeklyTodo.where(group_id: params[:group_id], creation_week: Date.parse(params[:date]).at_beginning_of_week)
+        return true if todo.present?
+
+        WeeklyTodo.create(group_id: params[:group_id], creation_week: Date.parse(params[:date]).at_beginning_of_week)
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/weekly_todo_controller.rb:48]: create should not happen on create

1. also should not be able to create or update records, created of last week or earlier

File: devsnest/spec/requests/weekly_todo_spec.rb [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        body = JSON.parse(response.body, symbolize_names: true)
+        expect(body[:data][:attributes][:sheet_filled]).to eq(parameters_update[:data][:attributes][:sheet_filled])
+        expect(body[:data][:attributes][:batch_leader_rating]).to eq(parameters_update[:data][:attributes][:batch_leader_rating])
+        expect(body[:data][:attributes][:extra_activity]).to eq(parameters_update[:data][:attributes][:extra_activity])
+        expect(body[:data][:attributes][:group_activity_rating]).to eq(parameters_update[:data][:attributes][:group_activity_rating])
+        expect(body[:data][:attributes][:moral_status]).to eq(parameters_update[:data][:attributes][:moral_status])
+        expect(body[:data][:attributes][:obstacles]).to eq(parameters_update[:data][:attributes][:obstacles])
+        expect(body[:data][:attributes][:comments]).to eq(parameters_update[:data][:attributes][:comments])
+        expect(body[:data][:attributes][:todo_list]).to eq(parameters_update[:data][:attributes][:todo_list])
+      end
```
Comment (vedansh) [inline devsnest/spec/requests/weekly_todo_spec.rb:212]: new line

File: devsnest/spec/requests/weekly_todo_spec.rb [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+
+    context ' Authoize Create ' do
+      before(:each) do
+        WeeklyTodo.last.destroy
+      end
+      it 'retruns error when user does not belongs to the group and nither an admin nor the group batchleader' do
+        sign_in(user2)
+        post '/api/v1/weekly-todo', params: parameters_create.to_json, headers: HEADERS
+        expect(response).to have_http_status(400)
+      end
```
Comment (vedansh) [inline devsnest/spec/requests/weekly_todo_spec.rb:101]: new line missing, after every spec

**Review by LakshitXD — APPROVED** [review]

**Pull request merged**